NVorbis/VorbisReader.cs

```diff
@@ -1,249 +1,282 @@
-            {
-                throw new InvalidDataException("Container did not initialize!");
+                throw new ArgumentException("Could not load the specified container!", nameof(containerReader));
             }
+            _closeOnDispose = closeOnDispose;
             _containerReader = containerReader;
-
-            if (_decoders.Count == 0) throw new InvalidDataException("No Vorbis data found!");
+            _streamDecoder = _decoders[0];
         }
 
-        public VorbisReader(IPacketProvider packetProvider)
-            : this()
-        {
-            var ea = new NewStreamEventArgs(packetProvider);
-            NewStream(this, ea);
-            if (ea.IgnoreStream) throw new InvalidDataException("No Vorbis data found!");
-        }
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+        [Obsolete("Use \"new StreamDecoder(Contracts.IPacketProvider)\" and the container's NewStreamCallback or Streams property instead.", true)]
+        public VorbisReader(Contracts.IContainerReader containerReader) => throw new NotSupportedException();
 
-        bool LoadContainer(IContainerReader containerReader)
-        {
-            containerReader.NewStream += NewStream;
-            if (!containerReader.Init())
-            {
-                containerReader.NewStream -= NewStream;
-                return false;
-            }
-            return true;
-        }
+        [Obsolete("Use \"new StreamDecoder(Contracts.IPacketProvider)\" instead.", true)]
+        public VorbisReader(Contracts.IPacketProvider packetProvider) => throw new NotSupportedException();
+#pragma warning restore CS1591 // Missing XML comment for publicly visible type or member
 
-        void NewStream(object sender, NewStreamEventArgs ea)
+        private bool ProcessNewStream(Contracts.IPacketProvider packetProvider)
         {
-            var packetProvider = ea.PacketProvider;
-            var decoder = new VorbisStreamDecoder(packetProvider);
-            if (decoder.TryInit())
+            var decoder = CreateStreamDecoder(packetProvider);
+            decoder.ClipSamples = true;
+
+            var ea = new NewStreamEventArgs(decoder);
+            NewStream?.Invoke(this, ea);
+            if (!ea.IgnoreStream)
             {
                 _decoders.Add(decoder);
-                _serials.Add(packetProvider.StreamSerial);
-            }
-            else
-            {
-                // This is almost certainly not a Vorbis stream
-                ea.IgnoreStream = true;
+                return true;
             }
+            return false;
         }
 
+        /// <summary>
+        /// Cleans up this instance.
+        /// </summary>
         public void Dispose()
         {
             if (_decoders != null)
             {
                 foreach (var decoder in _decoders)
                 {
-                    decoder.Dispose();
+                    (decoder as IDisposable)?.Dispose();
```

Comment:
What's the purpose of casting to IDisposable here? IStreamDecoder inherits from IDisposable right?

Reply:
At one point during development I'd removed IDisposable from the IStreamDecoder interface.  I later added it back in (I believe while working through the IPacketProvider lifetime handling).  It still works fine as-is, but for cleanliness I'll change it to call directly in the next push.

Reply:
It's up to you. I don't want to waste your time with small cleanliness issues. Unless you want me to keep pointing this kind of stuff out, I was planning on stopping. ReSharper nags me about stuff like this all the time so it's easy for me to do the same. Also, sorry if this review is taking me a while. I've been busy with stuff, but I've gotten through most of the files.

Reply:
No worries.  You may notice I'm also not in any hurry to reply. :)

Do as much (or little) review as you feel is necessary.  I'm good with whatever you do there.